src/kvstore.c

```diff
@@ -629,24 +629,38 @@ unsigned long kvstoreDictSize(kvstore *kvs, int didx)
     return dictSize(d);
 }
 
-dictIterator *kvstoreDictGetIterator(kvstore *kvs, int didx)
+kvstoreDictIterator *kvstoreGetDictIterator(kvstore *kvs, int didx)
 {
-    dict *d = kvstoreGetDict(kvs, didx);
-    return dictGetIterator(d);
+    kvstoreDictIterator *kvs_di = zmalloc(sizeof(*kvs_di));
+    kvs_di->kvs = kvs;
+    kvs_di->didx = didx;
+    dictInitIterator(&kvs_di->di, kvstoreGetDict(kvs, didx));
+    return kvs_di;
 }
 
-dictIterator *kvstoreDictGetSafeIterator(kvstore *kvs, int didx)
+kvstoreDictIterator *kvstoreGetDictSafeIterator(kvstore *kvs, int didx)
 {
-    dict *d = kvstoreGetDict(kvs, didx);
-    return dictGetSafeIterator(d);
+    kvstoreDictIterator *kvs_di = zmalloc(sizeof(*kvs_di));
+    kvs_di->kvs = kvs;
+    kvs_di->didx = didx;
+    dictInitSafeIterator(&kvs_di->di, kvstoreGetDict(kvs, didx));
+    return kvs_di;
 }
 
-/* Reset the iterator and free the iterator. */
-void kvstoreDictReleaseIterator(kvstore *kvs, int didx, dictIterator *iter)
+/* Free the kvs_di returned by kvstoreGetDictIterator and kvstoreGetDictSafeIterator. */
+void kvstoreReleaseDictIterator(kvstoreDictIterator *kvs_di)
 {
     /* The dict may be deleted during the iteration process, so here need to check for NULL. */
-    if (kvstoreDictExists(kvs, didx)) dictResetIterator(iter);
-    dictFreeIterator(iter);
+    if (kvstoreGetDict(kvs_di->kvs, kvs_di->didx)) dictResetIterator(&kvs_di->di);
+
+    zfree(kvs_di);
+}
+
+/* Get the next element of the dict through kvstoreDictIterator and dictNext. */
+dictEntry *kvstoreDictIteratorNext(kvstoreDictIterator *kvs_di)
+{
+    dictEntry *de = kvs_di->di.d ? dictNext(&kvs_di->di) : NULL;
```

Comment:
maybe we need to call `kvstoreGetDict` instead of `kvs_di->di.d`?
this will make it unnecessary to have a special fix in pubsubShardUnsubscribeAllChannelsInSlot, right?

Reply:
right! i updated it